src/osemgrep/cli_install_semgrep_pro/Install_semgrep_pro_subcommand.ml

```diff
@@ -49,24 +48,40 @@ let add_semgrep_pro_version_stamp current_executable_path =
   (* THINK: does this append or write entirely? *)
   UFile.write_file pro_version_stamp_path Version.version
 
-let download_semgrep_pro (caps : < Cap.network ; .. >) platform_kind dest =
+let download_semgrep_pro_async (caps : < Cap.network ; .. >) platform_kind dest
+    =
   let dest = !!dest in
   match (Semgrep_settings.load ()).api_token with
   | None ->
       Logs.err (fun m ->
           m "No API token found, please run `semgrep login` first.");
-      false
+      Lwt.return_false
   | Some token -> (
       let caps = Auth.cap_token_and_network token caps in
-      match Semgrep_App.fetch_pro_binary caps platform_kind with
-      | Error (_, { code = 401; _ }) ->
+      match%lwt Semgrep_App.fetch_pro_binary caps platform_kind with
+      | Ok { body = Ok body; response; _ } ->
+          (* Make sure no such binary exists. We have had weird situations
+           * when the downloaded binary was corrupted, and overwriting it did
+           * not fix it, but it was necessary to `rm -f` it.
+           *)
+          if Sys.file_exists dest then FileUtil.rm [ dest ];
+
+          (* TODO: does this matter if we don't have a progress bar? *)
+          let _file_size =
+            Cohttp.(Header.get (Response.headers response) "Content-Length")
+            |> Option.map int_of_string |> Option.value ~default:0
+          in
+
+          UFile.write_file (Fpath.v dest) body;
```

Comment:
If we don't shadow `dest` above then we don't have to `re-make` a path here.

Reply:
See other replies in this file